[RFC] ath79/ar71xx: Add Support for the TP-LINK CPE220 V3.0 Device

apcameron opened the pull request:
This adds support for a popular low-cost 2.4GHz N based AP
Specifications:
SoC: Qualcomm Atheros QCA9533 (650MHz)
RAM: 64MB
Storage: 8 MB SPI NOR
Wireless: 2.4GHz N based built into SoC 2x2
Ethernet: 2x 100/10 Mbps, integrated into SoC
Installation:
Flash factory image through stock firmware WEB UI
or through TFTP
To get to TFTP recovery just hold reset button while powering on for
around 4-5 seconds and release.
Rename factory image to recovery.bin
Stock TFTP server IP:192.168.0.100
Stock device TFTP adress:192.168.0.254

Signed-off-by: Andrew Cameron apcameron@softhome.net

Dmesg files from both Branches
[dmesg_ar71xx.txt](https://github.com/openwrt/openwrt/files/3759370/dmesg_ar71xx.txt)
[dmesg_ath79.txt](https://github.com/openwrt/openwrt/files/3759371/dmesg_ath79.txt)

<del>NOTE- The device functions correctly on the ATH71XX branch.
The ATH79 Branch suffers from the same bug that the TP-LINK CPE210 V2 and TP-LINK CPE210 V3 have where the WIFI is Deaf  See BUG ID [FS#2333](https://bugs.openwrt.org/index.php?do=details&task_id=2333&string=cpe210&type%5B0%5D=&sev%5B0%5D=&pri%5B0%5D=&due%5B0%5D=&reported%5B0%5D=&cat%5B0%5D=&status%5B0%5D=open&percent%5B0%5D=&opened=&dev=&closed=&duedatefrom=&duedateto=&changedfrom=&changedto=&openedfrom=&openedto=&closedfrom=&closedto=) and [FS#2478](https://bugs.openwrt.org/index.php?do=details&task_id=2478&string=cpe210&type%5B0%5D=&sev%5B0%5D=&pri%5B0%5D=&due%5B0%5D=&reported%5B0%5D=&cat%5B0%5D=&status%5B0%5D=open&percent%5B0%5D=&opened=&dev=&closed=&duedatefrom=&duedateto=&changedfrom=&changedto=&openedfrom=&openedto=&closedfrom=&closedto=) at https://bugs.openwrt.org/</del>

adrianschmutzler commented:
Please put ar71xx and ath79 into separate commits. Keep in mind that the chances for adding device support in ar71xx are close to zero. Even ath79 support will most probably not be backported, but only added to master.

adrianschmutzler commented on target/linux/ath79/base-files/etc/board.d/01_leds:
```diff
@@ -218,6 +218,15 @@ tplink,cpe510-v3)
 	ucidef_set_led_rssi "rssihigh" "RSSIHIGH" "tp-link:green:link4" "wlan0" "76" "100" "-75" "13"
 	;;
+tplink,cpe220-v3)
+	ucidef_set_led_netdev "lan" "LAN" "tp-link:green:lan" "eth1"
```
Any reason why you do not label this "lan0" and merge it with definitions for cpe210-v1?

adrianschmutzler commented on target/linux/ath79/dts/qca9533_tplink_cpe220-v3.dts:
```diff
@@ -0,0 +1,151 @@
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+#include "qca953x.dtsi"
```
One should create a common dtsi from this one and qca9533_tplink_cpe210.dtsi.

The only differences are eth1 and the lan leds. I would call the parent dtsi
qca9533_tplink_cpe.dtsi
or
qca9533_tplink_cpexxx.dtsi

adrianschmutzler commented on target/linux/ath79/dts/qca9533_tplink_cpe220-v3.dts:
```diff
@@ -0,0 +1,151 @@
+
+	keys {
+		compatible = "gpio-keys-polled";
+		poll-interval = <100>;
```
Please change to gpio-keys.

adrianschmutzler commented on target/linux/ath79/image/generic-tp-link.mk:
```diff
@@ -240,6 +240,19 @@ define Device/tplink_cpe220-v2
+  DEVICE_PACKAGES := rssileds
+  TPLINK_BOARD_ID := CPE220V3
+  LOADER_TYPE := elf
+  SUPPORTED_DEVICES += cpe220-v3
```
This is only needed in case ar71xx support is added. I would drop it for now.

apcameron commented:
@adrianschmutzler The Primary Purpose of the PR is to try to get the WIFI Bug resolved for the CPE210 V2 and CPE210 V3 as well as this device on the ath79 branch which you are well aware of hence including the code for both ar71xx where it works and ath79 where it does not work.
I did try to create a common dtsi file but it did not work and I could not access the device on its ethernet port.
I will relook at it in the next few days when I have more time and work on splitting the commits.

apcameron committed:
ath79: add Support for the TP-LINK CPE220 V3.0 Device

This adds support for a popular low-cost 2.4GHz N based AP

Specifications:
SoC: Qualcomm Atheros QCA9533 (650MHz)
RAM: 64MB
Storage: 8 MB SPI NOR
Wireless: 2.4GHz N based built into SoC 2x2
Ethernet: 2x 100/10 Mbps, integrated into SoC

Installation:
Flash factory image through stock firmware WEB UI
or through TFTP
To get to TFTP recovery just hold reset button while powering on for
around 4-5 seconds and release.
Rename factory image to recovery.bin
Stock TFTP server IP:192.168.0.100
Stock device TFTP adress:192.168.0.254

Signed-off-by: Andrew Cameron apcameron@softhome.net

apcameron committed:
ar71xx: add Support for the TP-LINK CPE220 V3.0 Device

This adds support for a popular low-cost 2.4GHz N based AP

Specifications:
SoC: Qualcomm Atheros QCA9533 (650MHz)
RAM: 64MB
Storage: 8 MB SPI NOR
Wireless: 2.4GHz N based built into SoC 2x2
Ethernet: 2x 100/10 Mbps, integrated into SoC

Installation:
Flash factory image through stock firmware WEB UI
or through TFTP
To get to TFTP recovery just hold reset button while powering on for
around 4-5 seconds and release.
Rename factory image to recovery.bin
Stock TFTP server IP:192.168.0.100
Stock device TFTP adress:192.168.0.254

Signed-off-by: Andrew Cameron apcameron@softhome.net

apcameron commented:
@adrianschmutzler I have split it into two commits as requested and created common qca9533_tplink_cpexxx.dtsi file. NOTE: The ar71xx commit depends on the ath79 commit for the tools/firmware-utils/src/tplink-safeloader.c updates

adrianschmutzler commented on target/linux/ath79/dts/qca9533_tplink_cpe210.dtsi:
```diff
@@ -4,20 +4,9 @@
 #include <dt-bindings/gpio/gpio.h>
 #include <dt-bindings/input/input.h>
```
You can remove those includes, as they are already in the parent DTSI.

apcameron replied:
I will remove this

adrianschmutzler commented on target/linux/ath79/dts/qca9533_tplink_cpe210.dtsi:
```diff
@@ -4,20 +4,9 @@
-		label-mac-device = &eth0;
-	};
 
 	leds {
```
I don't think this will work, as leds is already defined in parent DTSI. You will need to add a label to parent leds
`leds: leds {`
and then use `&leds` outside of the root node to add additional leds in this file (as you do for eth0 etc.).

apcameron replied:
This compiles cleanly and works when tested on my CPE210 V2 and CPE220 V3

adrianschmutzler commented on target/linux/ath79/dts/qca9533_tplink_cpe210.dtsi:
```diff
@@ -27,94 +16,6 @@
-
-	flash@0 {
-		#address-cells = <1>;
-		#size-cells = <1>;
```
Those two are removed in new parent DTSI?

apcameron replied:
I will add it back but do not believe it is needed

adrianschmutzler replied:
If you are _sure_ that it's not needed, you can remove it. Please add a brief comment about the remove in the commit message than. Otherwise, keep it as it will be a transparent copy/paste then.

apcameron replied:
I have added it back for now

adrianschmutzler commented on target/linux/ath79/dts/qca9533_tplink_cpexxx.dtsi:
```diff
@@ -0,0 +1,119 @@
+        mtd-cal-data = <&art 0x1000>;
+        mtd-mac-address = <&info 0x8>;
+};
+
```
Remove empty line at EOF.

adrianschmutzler commented on target/linux/ath79/dts/qca9533_tplink_cpexxx.dtsi:
```diff
@@ -0,0 +1,119 @@
+		led-failsafe = &system;
+		led-running = &system;
+		led-upgrade = &system;
+		label-mac-device = &eth0;
```
You use eth0 here, but only have wmac defined in the parent DTSI. Since eth0 and wmac are the same, it would be easiest to just use wmac for label-mac-device here (and add a short comment about the change in commit message).

apcameron replied:
This compiles cleanly and works when tested on my CPE210 V2 and CPE220 V3

adrianschmutzler replied:
Can you post the result of `ls /proc/device-tree/leds/` on one of the devices?
Does the led4 blink during boot?

The above is just to satisfy my curiosity. To my knowledge, OpenWrt consequently uses the node referencing scheme as I pointed out in my comments. So even if your version is working, I'd insist on using the &leds variant.

apcameron replied:
Yes led4 blinks during boot here is the output of the ls -al
drwxr-xr-x    8 root     root             0 Jan  1  1970 .
drwxr-xr-x   10 root     root             0 Jan  1  1970 ..
-r--r--r--    1 root     root            10 Oct 26 21:06 compatible
drwxr-xr-x    2 root     root             0 Oct 26 21:06 lan0
drwxr-xr-x    2 root     root             0 Oct 26 21:06 lan1
drwxr-xr-x    2 root     root             0 Oct 26 21:06 link1
drwxr-xr-x    2 root     root             0 Oct 26 21:06 link2
drwxr-xr-x    2 root     root             0 Oct 26 21:06 link3
drwxr-xr-x    2 root     root             0 Jan  1  1970 link4
-r--r--r--    1 root     root             5 Oct 26 21:06 name

adrianschmutzler commented on target/linux/ath79/dts/qca9533_tplink_cpe220-v3.dts:
```diff
@@ -0,0 +1,42 @@
+/dts-v1/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
```
Includes can be removed.

adrianschmutzler commented on target/linux/ath79/dts/qca9533_tplink_cpe220-v3.dts:
```diff
@@ -0,0 +1,42 @@
+	compatible = "tplink,cpe220-v3", "qca,qca9533"; 
+	model = "TP-Link CPE220 v3"; 
+
+	leds {
```
See my leds comment above. Same applies here (and you can also remove the gpio-leds, as it will be inherited from parent).

adrianschmutzler commented on target/linux/ath79/dts/qca9533_tplink_cpe220-v3.dts:
```diff
@@ -0,0 +1,42 @@
+	leds {
+		compatible = "gpio-leds";
+
+		lan {
```
Also use "lan0" here.

adrianschmutzler commented:
@apcameron 
Some formal issues first: The "Add" after ar71xx:/ath79: in commit title needs to be lower case.

Both your commit messages have an indent (most likely introduced by copying from git log?) which is unnecessary and should be removed. While at it, you might add some empty lines to separate "Installation"/specs etc.

The formal issues apply to both ath79 and ar71xx. My code review is limited to ath79.

apcameron commented:
@adrianschmutzler I have made the changes you requested and tested it again on my TP-LINK CPE220 v3 and TP-LINK CPE210 v2.  Now if someone could just solve problem of WIFI being Deaf.

adrianschmutzler commented:
@apcameron Thanks for doing the changes. Since you (if I understood correctly) want to keep this open until the WiFi problems are resolved, you might put the DTSI changes (qca9533_tplink_cpexxx.dtsi and qca9533_tplink_cpe210.dtsi) into a separate commit, so I may already pull them into master (without cpe220-v3.dts of course). This should make this PR a little more light-weight.

Concerning the WiFi issue, let's take the opportunity to collect what we know about it yet in a condensed form, so it can be discussed (maybe I will sent the aggregated info to the list then, too).
Please correct where I am missing things or am incorrect about something. I will add/correct stuff directly in this post.

1. Symptoms: WiFi is up, but no data is transferred over it (can you connect to an AP?)
2. Affected devices: TP-Link CPE210 v2/v3, CPE220 v3 (all QCA9533 rev. 2?); no other QCA9533 devices known to be affected (specific to CPE or to QCA9533 rev. 2?)
3. For a certain model, there are devices which are working correctly and others which don't. There is no known indicator to find out whether a device works or not. The state of a device does not change (always working or never working).
4. So far, only 2.4 GHZ-only devices were affected
5. There is no diagnostic output that indicates a WiFi problem. dmesg/logread look normal, there is no difference when compared between working and not-working devices (the interface is shown as up? Is there traffic shown?)
6. The problem seems to be present from the beginning (device support patches), it just has been overlooked since it's not occurring on every devices.
7. The ar71xx firmware for all devices works flawlessly, so it is an ath79-specific problem.

I currently own a CPE210v2 with the bug for testing if required.

IronicSven commented:
@adrianschmutzler another thing we know is that in comparison to ar71xx a gpio seems to be missing: https://forum.openwrt.org/t/ath79-tp-link-cpe210-v2-0-wifi-not-working/40666/18

apcameron commented:
Is there an easy way to add the GPIO below for testing as the ar71xx branch has it?
gpiochip1: GPIOs 494-511, ath9k-phy0:
 gpio-495 (                    |ath9k-phy0          ) in  lo

adrianschmutzler commented:
@apcameron I'm not sure. You might try a gpio-hog:
```
&gpio {
	test {
		gpio-hog;
		gpios = <495 GPIO_ACTIVE_LOW>;
		output-low;
		line-name = "tp-link:ext:test";
	};
```

I'm not sure about the correct high-low values there or whether it will be a good idea at all.

apcameron commented:
I have tried that and other ways but it did not work.
I still only see the following
-----
cat /sys/kernel/debug/gpio 
gpiochip0: GPIOs 0-31, parent: platform/18040000.gpio, 18040000.gpio:
 gpio-11  (                    |tp-link:green:lan0  ) out lo    
 gpio-12  (                    |tp-link:green:lan1  ) out hi    
 gpio-13  (                    |tp-link:green:link1 ) out hi    
 gpio-14  (                    |tp-link:green:link2 ) out hi    
 gpio-15  (                    |tp-link:green:link3 ) out hi    
 gpio-16  (                    |tp-link:green:link4 ) out hi    
 gpio-17  (                    |Reset button        ) in  hi IRQ
 gpio-20  (                    |sysfs               ) out lo    
-----
running a tcpdump in monitor mode shows that the software seems to be work but it seems that the Radio is turned off and is not transmitting or receiving.

adrianschmutzler commented:
I'm not really sure whether that gpio is the right direction to look at at all. Unfortunately, I'm no expert for this stuff, so I can't really help beyond that point. I think at the moment it would be best to finalize the problem overview I proposed and then send it to list.

adrianschmutzler commented:
Despite, maybe separate the DTS files as suggested earlier:

> Since you (if I understood correctly) want to keep this open until the WiFi problems are resolved, you might put the DTSI changes (qca9533_tplink_cpexxx.dtsi and qca9533_tplink_cpe210.dtsi) into a separate commit, so I may already pull them into master (without cpe220-v3.dts of course). This should make this PR a little more light-weight.

If you aren't keen on the authorship, I can also cut it out myself ;-)

apcameron commented:
If we have not progress in the next week of so I will put the dtsi changes in another commit. 
@blogic Do you have any idea why WIFI does not work on the TP LINK CPE210 V2/V3 and TP-LINK CPE220 V3 all based on the QCA9533-BL3A SOC. I saw one of your commits to get WIFI working on the ath79 branch https://git.openwrt.org/?p=openwrt/staging/blogic.git;a=commit;h=9423bb512e8815ad9598c62d41f23007118c0b16

adrianschmutzler commented:
I've just sent a mail to the list, too.

adrianschmutzler commented:
@apcameron Good news: blocktrron just provided a patch that fixes the issue on my CPE210v2:
https://github.com/openwrt/openwrt/commit/4c6fe32468bc60cc25a8c298498c0be3c73e7378

Please rebase your patches onto current master and test.

@bluelineXY If you still have the CPE210v3, please test with current master (I think yours worked from the beginning, so it would be interesting if it still works).

apcameron commented:
That's good news. I will let you know the results in the next day or so.

apcameron commented:
@adrianschmutzler @bluelineXY @robimarko I can confirm that WIFI is working now on the TP-LINK CPE220 V3 on the ath79 branch

adrianschmutzler commented:
@apcameron I will merge the ath79 patch (with some minor DTS adjustments) in the next hour. I won't merge the ar71xx patch (and don't think it will make much sense to wait on somebody else), but feel free to keep the PR open or close it based on your feeling about this.

apcameron commented:
@adrianschmutzler you can close the PR once the ath79 code has been merged. The ar71xx was to help with troubleshooting the wifi problems

adrianschmutzler commented:
ath79 has been merged. Closing without merging ar71xx as discussed above.
@apcameron Thanks for your endurance on that problem.

bluelineXY commented:
Hi, 
i have tested with the cpe210v3 and  it still works. 

Also i tested with a cpe210v2 and iw scan is now working. 

cool - great work !
